### PR TITLE
Simplified invokes: SDK-gen and program-gen implementation for dotnet and nodejs

### PR DESCRIPTION
**Pull request by Zaid-Ajaj**
# Description

This PR implements simplified invoke gen for both SDKs and PCL programs by allowing invokes to:
 - Accepts multi arguments as inputs (instead of a bag of properties)
 - Return a single value as output (instead of a bag of outputs)

Related to #7435 

This implementation handles dotnet and nodejs. Python and Go can be implementation on a separate pass. Currently both python and go will fail generation (sdk/program) when encountering simplified invokes

## Checklist

- [x] I have added tests that prove my fix is effective or that my feature works

- [x] I have run `make changelog` and committed the `changelog/pending/<file>` documenting my change

- [ ] Yes, there are changes in this PR that warrants bumping the Pulumi Service API version

### CODE REVIEW

**Comment by Zaid-Ajaj** [comment]
@iwahbe @Frassle @AaronFriel I've added you for a PR review if you have time 🙏 I also made a quick explanation of what the is and what it does, hope this helps 😄 


https://user-images.githubusercontent.com/13316248/204416456-dd202b75-1be6-410e-bee2-2925df5ef50d.mp4

File: pkg/codegen/nodejs/gen.go [hunk 1]
```diff
@@ -1422,9 +1515,11 @@ func (mod *modContext) getImportsForResource(member interface{}, externalImports
...
 			}
 		}
-		if member.Outputs != nil {
-			for _, p := range member.Outputs.Properties {
-				needsTypes = mod.getTypeImports(p.Type, false, externalImports, imports, seen) || needsTypes
+		if member.ReturnType != nil {
+			if objectType, ok := member.ReturnType.(*schema.ObjectType); ok && objectType != nil {
+				for _, p := range objectType.Properties {
+					needsTypes = mod.getTypeImports(p.Type, false, externalImports, imports, seen) || needsTypes
+				}
```
Comment (iwahbe) [inline pkg/codegen/nodejs/gen.go:1522]: Is there a reason we only get imports for object types? The benefit of having `member.ReturnType` be a simple `schema.Type` is less special casing.
```suggestion
			needsTypes = mod.getTypeImports(member.ReturnType, false, externalImports, imports, seen) || needsTypes
```

File: pkg/codegen/nodejs/gen.go [hunk 2]
```diff
@@ -1407,10 +1497,13 @@ func (mod *modContext) getImportsForResource(member interface{}, externalImports
...
-				for _, p := range method.Function.Outputs.Properties {
-					needsTypes =
-						mod.getTypeImportsForResource(p.Type, false, externalImports, imports, seen, res) || needsTypes
+
+			if method.Function.ReturnType != nil {
+				if objectType, ok := method.Function.ReturnType.(*schema.ObjectType); ok && objectType != nil {
+					for _, p := range objectType.Properties {
+						needsTypes =
+							mod.getTypeImportsForResource(p.Type, false, externalImports, imports, seen, res) || needsTypes
+					}
```
Comment (iwahbe) [inline pkg/codegen/nodejs/gen.go:1506]: Why do we need to special case this? See later comment.

File: pkg/codegen/nodejs/gen.go [hunk 3]
```diff
@@ -1184,6 +1228,15 @@ func (mod *modContext) genFunctionOutputVersion(
 
 	originalName := tokenToFunctionName(fun.Token)
 	fnOutput := fmt.Sprintf("%sOutput", originalName)
+	returnType := "void"
+	if fun.ReturnType != nil {
+		if objectType, ok := fun.ReturnType.(*schema.ObjectType); ok && objectType != nil {
```
Comment (iwahbe) [inline pkg/codegen/nodejs/gen.go:1233]: What if the return type is an `"$ref"` based object? We don't want to generate a special type there. I suspect we will need a `fun.GenerateReturnType() bool` to discriminate referenced objects from inline objects.

File: pkg/codegen/schema/schema.go [hunk 4]
```diff
@@ -1151,16 +1168,35 @@ func (pkg *Package) marshalFunction(f *Function) (FunctionSpec, error) {
 		outputs = &outs.ObjectTypeSpec
 	}
 
+	var returnType *ReturnTypeSpec
+	if f.ReturnType != nil {
+		returnType = &ReturnTypeSpec{}
+		if objectType, ok := f.ReturnType.(*ObjectType); ok {
```
Comment (iwahbe) [inline pkg/codegen/schema/schema.go:1174]: What about `"$ref"` based objects?

File: pkg/codegen/testing/test/program_driver.go [hunk 5]
```diff
@@ -104,6 +104,12 @@ var PulumiPulumiProgramTests = []ProgramTest{
 		Description: "Python regression test for #10914",
 		Skip:        allProgLanguages.Except("python"),
 	},
+	{
+		Directory:   "simplified-invokes",
+		Description: "Simplified invokes",
+		Skip:        allProgLanguages.Except("dotnet").Except("nodejs"),
```
Comment (iwahbe) [inline pkg/codegen/testing/test/program_driver.go:110]: This reads easier to me.
```suggestion
		Skip:        codegen.NewStringSet("python", "go")
```

File: pkg/codegen/testing/test/testdata/external-resource-schema/nodejs/argFunction.ts [hunk 6]
```diff
@@ -23,10 +22,10 @@ export interface ArgFunctionResult {
     readonly age?: number;
 }
 
+
 export function argFunctionOutput(args?: ArgFunctionOutputArgs, opts?: pulumi.InvokeOptions): pulumi.Output<ArgFunctionResult> {
-    return pulumi.output(args).apply(a => argFunction(a, opts))
+    return pulumi.output(args).apply((a: any) => argFunction(a, opts))
```
Comment (iwahbe) [inline pkg/codegen/testing/test/testdata/external-resource-schema/nodejs/argFunction.ts:27]: It would be really helpful if a change this big didn't change code for unchanged schemas! Could you apply this change to pu/pu and then rebase to cut down the number of files necessary to review here.

File: pkg/codegen/testing/test/testdata/output-funcs-tfbridge20/nodejs/getAmiIds.ts [hunk 7]
```diff
@@ -72,10 +71,14 @@ export interface GetAmiIdsResult {
     readonly sortAscending?: boolean;
 }
 
+/**
+ * Taken from pulumi-AWS to regress an issue
+ */
+/** @deprecated aws.getAmiIds has been deprecated in favor of aws.ec2.getAmiIds */
```
Comment (iwahbe) [inline pkg/codegen/testing/test/testdata/output-funcs-tfbridge20/nodejs/getAmiIds.ts:77]: Where did we depreciate this method? I don't see the code change that did this.

File: pkg/codegen/testing/test/testdata/simplified-invokes/schema.json [hunk 8]
```diff
@@ -0,0 +1,153 @@
+{
```
Comment (iwahbe) [inline pkg/codegen/testing/test/testdata/simplified-invokes/schema.json:1]: I would love to see a test for enums, objects, archives/assets and resources, both for positional arguments and as return types.

(I would be fine with forbidding returning resources, but that should be an explicit bind error, not a compilation error in generated code.)

**Review by iwahbe — CHANGES_REQUESTED** [review]
I left an initial set of feedback. It would be really helpful if you reverted the TS changes so that the only generated code that changes is the code we expect to see changed.

File: pkg/codegen/schema/bind.go [hunk 9]
```diff
@@ -1480,40 +1480,110 @@ func (t *types) bindFunctionDef(token string) (*Function, hcl.Diagnostics, error
...
+			outputs = outs
+		} else if spec.ReturnType.TypeSpec != nil {
+			out, outDiags, err := t.bindTypeSpec(path+"/outputs", *spec.ReturnType.TypeSpec, false)
+			diags = diags.Extend(outDiags)
+			if err != nil {
+				return nil, diags, fmt.Errorf("error binding outputs for function %v: %w", token, err)
+			}
+			returnType = out
+		} else {
+			// should not happen
```
[the remark this answers is not in the document]
Comment (Zaid-Ajaj) [reply pkg/codegen/schema/bind.go:1572]: Good catch, adding more docs and a better error if this happens here ✅

File: pkg/codegen/schema/schema.go [hunk 10]
```diff
@@ -538,8 +538,12 @@ type Function struct {
 	Comment string
 	// Inputs is the bag of input values for the function, if any.
 	Inputs *ObjectType
+	// Determines whether the input bag should be treated as a single argument or as multiple arguments.
```
[the remark this answers is not in the document]
Comment (Zaid-Ajaj) [reply pkg/codegen/schema/schema.go:541]: Yeah elaborated more on the docs here 👍

File: pkg/codegen/pcl/utilities.go [hunk 11]
```diff
@@ -162,3 +163,30 @@ func FixupPulumiPackageTokens(r *Resource) {
 		r.Token = "pulumi::" + name
 	}
 }
+
+// SortedFunctionParameters returns a list of properties of the input type from the schema
+// for an invoke function call which has multi argument inputs. We assume here
+// that the expression is an invoke which has it's args (2nd parameter) annotated
+// with the original schema type. The original schema type has properties sorted.
+// This is important because model.ObjectType has no guarantee of property order.
+func SortedFunctionParameters(expr *model.FunctionCallExpression) []*schema.Property {
```
[the remark this answers is not in the document]
Comment (Zaid-Ajaj) [reply pkg/codegen/pcl/utilities.go:172]: I changed the implementation to return an empty array instead of `nil` because if `MultiArgumentInputs` then the array of properties is must be non-empty which relies on invoke annotations. 

> Could we just spread multi-argument invoke parameters into the Signature.Paremeters field? Then retrieving the sorted parameters would be as simple as express.Signature.Parameters[1:].

That would change the shape of the invoke in PCL too and I am more inclined to leave as an object that corresponds with the schema inputs object.

File: pkg/codegen/nodejs/gen_program_expressions.go [hunk 12]
```diff
@@ -408,7 +452,16 @@ func (g *generator) GenFunctionCallExpression(w io.Writer, expr *model.FunctionC
...
 		g.Fprintf(w, "%s(", name)
 		if len(expr.Args) >= 2 {
-			g.Fgenf(w, "%.v", expr.Args[1])
+			if expr.Signature.MultiArgumentInputs {
+				arg, ok := expr.Args[1].(*model.ObjectConsExpression)
+				if ok {
+					g.genMultiArguments(w, arg, pcl.SortedFunctionParameters(expr))
+				} else {
+					g.Fgenf(w, "%.v", expr.Args[1])
+				}
```
[the remark this answers is not in the document]
Comment (Zaid-Ajaj) [reply pkg/codegen/nodejs/gen_program_expressions.go:461]: Yeah you are right, removed this case into the following:
```go
if expr.Signature.MultiArgumentInputs {
	args := expr.Args[1].(*model.ObjectConsExpression)
	g.genMultiArguments(w, args, pcl.SortedFunctionParameters(expr))
} else {
	g.Fgenf(w, "%.v", expr.Args[1])
}
```

File: pkg/codegen/dotnet/gen_program_expressions.go [hunk 13]
```diff
@@ -699,13 +703,63 @@ func resolvePropertyName(property string, overrides map[string]string) string {
...
 }
 
+// genMultiArguments takes the input bag (object) of a function invoke and spreads the values of that object
+// into multi-argument function call.
+// For example, { a: 1, b: 2 } with multiInputArguments: ["a", "b"] would become: 1, 2
+//
+// However, when positional optional parameters are omitted, then NULL is used where they should be.
+// Take for example { a: 1, c: 3 } with multiInputArguments: ["a", "b", "c"], it becomes 1, null, 3
+// because b was omitted and c was provided so b had to be NULL
+func (g *generator) genMultiArguments(w io.Writer, expr *model.ObjectConsExpression, multiArgs []*schema.Property) {
```
[the remark this answers is not in the document]
Comment (Zaid-Ajaj) [reply pkg/codegen/dotnet/gen_program_expressions.go:713]: Looking into this right now

File: pkg/codegen/schema/schema.go (hunk 4 above)
Comment (Zaid-Ajaj) [reply pkg/codegen/schema/schema.go:1174]: I _assumed_ that `$ref`-ed objects would be generated when they are part of the schema `types` but I will add some tests to make sure

File: pkg/codegen/testing/test/program_driver.go (hunk 5 above)
Comment (Zaid-Ajaj) [reply pkg/codegen/testing/test/program_driver.go:110]: Indeed

File: pkg/codegen/testing/test/testdata/external-resource-schema/nodejs/argFunction.ts (hunk 6 above)
Comment (Zaid-Ajaj) [reply pkg/codegen/testing/test/testdata/external-resource-schema/nodejs/argFunction.ts:27]: Yeah I've opened #11511 for this

File: pkg/codegen/testing/test/testdata/output-funcs-tfbridge20/nodejs/getAmiIds.ts (hunk 7 above)
Comment (Zaid-Ajaj) [reply pkg/codegen/testing/test/testdata/output-funcs-tfbridge20/nodejs/getAmiIds.ts:77]: This is because output-versioned invokes didn't have doc comments on them which I fixed in this PR, but it's now part of #11511 so it will go away from here

File: pkg/codegen/testing/test/testdata/simplified-invokes/schema.json (hunk 8 above)
Comment (Zaid-Ajaj) [reply pkg/codegen/testing/test/testdata/simplified-invokes/schema.json:1]: I will add some tests ✅

File: pkg/codegen/nodejs/gen.go (hunk 1 above)
Comment (Zaid-Ajaj) [reply pkg/codegen/nodejs/gen.go:1522]: I refactored it like this so it doesn't change any old behaviour which relied on `Outputs` before but I guess it is worth simplifying now

Comment (Zaid-Ajaj) [reply pkg/codegen/nodejs/gen.go:1522]: Actually applying this change results in the following redundant imports being added:
```ts
import * as inputs from "types/inputs";
import * as outputs from "types/inputs";
```
So leaving this one as is for now

File: pkg/codegen/nodejs/gen_program_expressions.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (Zaid-Ajaj) [reply pkg/codegen/nodejs/gen_program_expressions.go:461]: Actually found a bug with this one that I otherwise didn't catch before 😄

File: pkg/codegen/schema/schema.go (hunk 4 above)
Comment (iwahbe) [reply pkg/codegen/schema/schema.go:1174]: My concern here was that they will be deserialized both as part of the return type and in the tyes section. I doesn't look like `"$ref"` objects will be left as references on the way out.

File: pkg/codegen/schema/schema.go [hunk 14]
```diff
@@ -1515,15 +1551,57 @@ type ResourceSpec struct {
 	Methods map[string]string `json:"methods,omitempty" yaml:"methods,omitempty"`
 }
 
+// ReturnTypeSpec is either ObjectTypeSpec or TypeSpec
+type ReturnTypeSpec struct {
+	ObjectTypeSpec *ObjectTypeSpec
+	TypeSpec       *TypeSpec
+}
+
+func (returnTypeSpec *ReturnTypeSpec) UnmarshalJSON(inputJSON []byte) error {
```
Comment (iwahbe) [inline pkg/codegen/schema/schema.go:1560]: Unfortunately, we also need an `UnmarshalYAML`.

File: pkg/codegen/schema/schema.go [hunk 15]
```diff
@@ -538,8 +538,12 @@ type Function struct {
 	Comment string
 	// Inputs is the bag of input values for the function, if any.
 	Inputs *ObjectType
+	// Determines whether the input bag should be treated as a single argument or as multiple arguments.
+	MultiArgumentInputs bool
 	// Outputs is the bag of output values for the function, if any.
 	Outputs *ObjectType
+	// The return type of the function, if any.
+	ReturnType Type
```
Comment (Frassle) [inline pkg/codegen/schema/schema.go:546]: I'm surprised this is just being add as a `Type`, how would we mark the single return value as secret?

Comment (Zaid-Ajaj) [reply pkg/codegen/schema/schema.go:546]: @Frassle in this iteration, this is not yet possible.

File: pkg/codegen/schema/schema.go (hunk 4 above)
Comment (Zaid-Ajaj) [reply pkg/codegen/schema/schema.go:1174]: There should definitely be a distinction: `$ref` types should be generated as they are encountered. 

My tests are showing at least the dotnet SDK-gen that:
  - Object type definitions in `types` are _not_ generated separately
  - Referenced object type definitions from the ReturnType are generated as they are encountered (should not be the case)

File: pkg/codegen/schema/schema.go (hunk 15 above)
Comment (Frassle) [reply pkg/codegen/schema/schema.go:546]: Is it going to be harder to add support for that in a later iteration while also supporting Type at this location in the schema? I would have made a return parameter nested object for this, which in this iteration simply has one field type. But if y'all think something like that will be easy to add later with this as it is then fine.

Comment (Zaid-Ajaj) [reply pkg/codegen/schema/schema.go:546]: @Frassle it can be added as a field to the `Function` definition, not necessarily the return type itself.

Comment (Frassle) [reply pkg/codegen/schema/schema.go:546]: Well yes I suppose it could but I'd find that distasteful. But being a question of taste and given I don't work much with this stuff I'll just leave it at that.

Comment (Zaid-Ajaj) [reply pkg/codegen/schema/schema.go:546]: @Frassle it's the same for `PropertySpec`, the secret-ness is not a field of the `Type` of the property but adjacent to it (as a field to the `PropertySpec` itself). Adding `sercret: bool` to `FunctionSpec` would follow the same pattern

File: pkg/codegen/schema/schema.go (hunk 14 above)
Comment (Zaid-Ajaj) [reply pkg/codegen/schema/schema.go:1560]: added the implementation, similar to UnmarshalJSON

File: pkg/codegen/schema/schema.go (hunk 4 above)
Comment (Zaid-Ajaj) [reply pkg/codegen/schema/schema.go:1174]: Fixed this issue BTW ✅

File: pkg/codegen/testing/test/program_driver.go [hunk 16]
```diff
@@ -97,6 +97,12 @@ var PulumiPulumiProgramTests = []ProgramTest{
 		Description: "Python regression test for #10914",
 		Skip:        allProgLanguages.Except("python"),
 	},
+	{
+		Directory:   "simplified-invokes",
+		Description: "Simplified invokes",
+		Skip:        codegen.NewStringSet("python", "go"),
+		SkipCompile: codegen.NewStringSet("dotnet", "nodejs"),
```
Comment (iwahbe) [inline pkg/codegen/testing/test/program_driver.go:104]: I'm assuming that compiling is blocked on https://github.com/pulumi/pulumi/pull/11564. Is that right?

**Review by iwahbe — COMMENTED** [review]
The changes look good to me.

Can you remove whatever you did that caused whitespace changes to all generated TS functions?

File: pkg/codegen/testing/test/program_driver.go (hunk 16 above)
Comment (Zaid-Ajaj) [reply pkg/codegen/testing/test/program_driver.go:104]: Yeah waiting for that to be merged then gonna rebase this on top ✅

Comment (Zaid-Ajaj) [reply pkg/codegen/testing/test/program_driver.go:104]: I stand corrected. This is actually skipping program-gen compilation because `Pulumi.Std` and `@pulumi/std` are not real packages (yet). The sdk-gen already compiles both dotnet and nodejs, that's why it was compiling them (because it was only skipping compiling programs)

That said, I think it would be great to be able to test with the generated SDK by a project reference since we have a schema to go from but I don't think that's possible right now 🤔

**Comment by Zaid-Ajaj** [comment]
@iwahbe @Frassle I updated the PR such that it no longer creates diffs in unaffected TS files.

File: pkg/codegen/testing/test/testdata/output-funcs/nodejs/funcWithEmptyOutputs.ts [hunk 17]
```diff
@@ -7,7 +7,7 @@ import * as utilities from "./utilities";
 /**
  * n/a
  */
-export function funcWithEmptyOutputs(args: FuncWithEmptyOutputsArgs, opts?: pulumi.InvokeOptions): Promise<FuncWithEmptyOutputsResult> {
+export function funcWithEmptyOutputs(args: FuncWithEmptyOutputsArgs, opts?: pulumi.InvokeOptions): Promise<void> {
```
Comment (iwahbe) [inline pkg/codegen/testing/test/testdata/output-funcs/nodejs/funcWithEmptyOutputs.ts:10]: This is a breaking change. While I think the resulting code is better, I would do this in a separate PR so it is independently reviewable and revertible.

File: pkg/codegen/nodejs/gen.go [hunk 18]
```diff
@@ -1102,27 +1157,55 @@ func (mod *modContext) genFunction(w io.Writer, fun *schema.Function) (functionF
...
+				fmt.Fprintf(w, "%s?: ", prop.Name)
+				// since we already applied the ? to the type, we can simplify
+				// the optional-ness of the type
+				switch propType := prop.Type.(type) {
+				case *schema.OptionalType:
+					fmt.Fprintf(w, "%s, ", mod.typeString(propType.ElementType, false, nil))
+				default:
+					fmt.Fprintf(w, "%s, ", mod.typeString(propType, false, nil))
+				}
+			}
```
Comment (iwahbe) [inline pkg/codegen/nodejs/gen.go:1179]: `IsRequired` is defined as:

https://github.com/pulumi/pulumi/blob/66db80ced5dd0e831f70735b777b4be335068bb2/pkg/codegen/schema/schema.go#L359-L363

We don't need the full type switch here, since we know the underlying type:

```suggestion
			if prop.IsRequired() {
				fmt.Fprintf(w, "%s: ", prop.Name)
				fmt.Fprintf(w, "%s, ", mod.typeString(prop.Type, false, nil))
			} else {
				fmt.Fprintf(w, "%s?: ", prop.Name)
				// since we already applied the ? to the type, we can simplify
				// the optional-ness of the type
				fmt.Fprintf(w, "%s, ", mod.typeString(propType.(*schema.OptionalType).ElementType, false, nil))
				}
			}
```

File: pkg/codegen/nodejs/gen.go [hunk 19]
```diff
@@ -1102,27 +1157,55 @@ func (mod *modContext) genFunction(w io.Writer, fun *schema.Function) (functionF
...
+	invokeCall := runtimeInvokeFunction(fun)
 	// Now simply invoke the runtime function with the arguments, returning the results.
-	fmt.Fprintf(w, "    return pulumi.runtime.invoke(\"%s\", {\n", fun.Token)
+	fmt.Fprintf(w, "    return pulumi.runtime.%s(\"%s\", {\n", invokeCall, fun.Token)
 	if fun.Inputs != nil {
 		for _, p := range fun.Inputs.Properties {
 			// Pass the argument to the invocation.
 			body := fmt.Sprintf("args.%s", p.Name)
+			if fun.MultiArgumentInputs {
+				body = fmt.Sprintf("%s", p.Name)
```
Comment (iwahbe) [inline pkg/codegen/nodejs/gen.go:1206]: Why are we doing this instead of just assigning?
```suggestion
				body = p.Name
```

File: pkg/codegen/schema/schema.go [hunk 20]
```diff
@@ -1539,6 +1641,152 @@ type FunctionSpec struct {
...
+	}
+
+	if objectData == nil {
+		return true, nil
+	}
+
+	return len(*objectData) == 0, nil
+}
+
+func UnmarshalFunctionSpec(funcSpec *FunctionSpec, data map[string]RawMessage) error {
```
Comment (iwahbe) [inline pkg/codegen/schema/schema.go:1657]: Why do we need to custom unmarshal the entire `FunctionSpec`? This seems really error prone.

Also, does this need to be public?

File: pkg/codegen/schema/schema.go [hunk 21]
```diff
@@ -1522,15 +1558,81 @@ type ResourceSpec struct {
...
+			return err
+		}
+	}
+
+	returnTypeSpec.TypeSpec = typeSpec
+	returnTypeSpec.ObjectTypeSpec = objectSpec
+	return nil
+}
+
+func (returnTypeSpec *ReturnTypeSpec) UnmarshalYAML(node *yaml.Node) error {
```
Comment (iwahbe) [inline pkg/codegen/schema/schema.go:1594]: TIL this could take a `*yaml.Node`. 😄 

I think this function is identical to `UnmarshalJSON`. Could we set this up so we pass in a decoder function and do some type casting to avoid the duplication. Having two functions that should be the same module type seems error prone.

**Review by iwahbe — CHANGES_REQUESTED** [review]
I have a couple of suggestions for code quality. My last area of concern is serialization, but I think that should be relatively quick to fix. I'm hoping there is a way to do it without manual re-implementation.

File: pkg/codegen/schema/schema.go (hunk 20 above)
Comment (Zaid-Ajaj) [reply pkg/codegen/schema/schema.go:1645]: Because I need  to control how two fields are unmarshalled: `Outputs` and `ReturnType`, the former being derived from the latter. Same goes for marshalling. However, it does not need to be public. I will change that

File: pkg/codegen/schema/schema.go (hunk 21 above)
Comment (Zaid-Ajaj) [reply pkg/codegen/schema/schema.go:1582]: > TIL this could take a *yaml.Node. 😄

Stole that from the implementation of `RawMessage` 😁 

So after a bit of testing, it looks like `UnmarshalYAML` is broken, i.e. it is not being called and doesn't work however I try. Changing the signature to `func (funcSpec *FunctionSpec) UnmarshalYAML(input []byte) error`yields the same results 😞

Comment (Zaid-Ajaj) [reply pkg/codegen/schema/schema.go]: Didn't need a to type case anything, a decoder function did the trick and added tests to go from and to JSON/YAML
```go
type Decoder func([]byte, interface{}) error

func (returnTypeSpec *ReturnTypeSpec) UnmarshalReturnTypeSpec(data []byte, decode Decoder) error {
   // implementation
}

func (returnTypeSpec *ReturnTypeSpec) UnmarshalJSON(inputJSON []byte) error {
	return returnTypeSpec.UnmarshalReturnTypeSpec(inputJSON, json.Unmarshal)
}

func (returnTypeSpec *ReturnTypeSpec) UnmarshalYAML(inputYAML []byte) error {
	return returnTypeSpec.UnmarshalReturnTypeSpec(inputYAML, yaml.Unmarshal)
}
```

File: pkg/codegen/dotnet/gen.go [hunk 22]
```diff
@@ -1336,16 +1351,67 @@ func allOptionalInputs(fun *schema.Function) bool {
...
+	// If the function has an object return type, it is also a normal invoke function.
+	// because the deserialization can handle it
+	if _, isMap := fun.ReturnType.(*schema.MapType); isMap {
+		return "Invoke"
+	}
+
+	// Anything else needs to be handled by InvokeSingle
+	// which expects an object with a single property to be returned
+	// then unwraps the value from that property
+	return "InvokeSingle"
```
[the remark this answers is not in the document]
Comment (Zaid-Ajaj) [reply pkg/codegen/dotnet/gen.go]: Done ✅

File: pkg/codegen/schema/schema.go (hunk 20 above)
Comment (iwahbe) [reply pkg/codegen/schema/schema.go]: Does this need to be done at the `FunctionSpec` level. I imagine it would be much easier and less error prone to do at the bind level. 

P.S. The current implementation directly calls `json.Unmarshal`, which won't work for yaml.

Comment (Zaid-Ajaj) [reply pkg/codegen/schema/schema.go]: > Does this need to be done at the FunctionSpec level. I imagine it would be much easier and less error prone to do at the bind level.

I tried removing the unmarshalling logic and instead only have custom marshalling for `ReturnTypeSpec`, however that doesn't work with YAML because if I put the attribute `yaml:"outputs"` on the `ReturnTypeSpec`, then it fails saying that there are duplicate `outputs` definitions because the `Outputs` fields is still there. It can be simplified when we remove `Outputs` entirely

> P.S. The current implementation directly calls json.Unmarshal, which won't work for yaml.

It works because it uses `map[string]RawMessage` as an intermediate data structure for both JSON and YAML. See unit tests I've added in `schema_test.go` for marshalling and unmarshalling `FunctionSpec`

**Review by iwahbe — APPROVED** [review]
I'm satisfied. LGTM!

**Comment by Zaid-Ajaj** [comment]
bors merge
